tls/autotls-client.md

```diff
@@ -0,0 +1,396 @@
+	```json
+	{
+      "identifiers": [
+        {
+          "type": "dns",
+          "value": "*.{b36peerid}.libp2p.direct"
+        }
+      ]
+	}
+	```
+7. From the ACME server response, the node MUST save the entry with `type` of `dns-01` and derive [`keyAuthorization`](https://datatracker.ietf.org/doc/html/rfc8555#section-8.1) from that.
+8. From the ACME server response's `dns-01` field, the node MUST also save the value on the `url` field of the JSON body, here called `chalUrl`. This is used in the ACME signalling phase.
+9. From the ACME server response's, the node MUST also save the value on the `location` header, here called `orderUrl`. This is used in the ACME signalling phase.
+10. From the ACME server response's, the node MUST also save the value on the `finalize` field of the JSON body, here called `finalizeUrl`. This is used in the ACME signalling phase.
+
+
+
+
+## Sending challenge to AutoTLS broker
+1. The node sends `keyAuthorization` to the AutoTLS broker (e.g. `registration.libp2p.direct`). This requires a [peer ID authentication](https://github.com/libp2p/specs/blob/master/http/peer-id-auth.md) between node and broker:
+	1. Node sends GET request to the AutoTLS broker's `/v1/_acme-challenge` endpoint and extracts `challenge-node`, `public-key` and `opaque` from the `www-authenticate` response header.
+	2. Node generates 32-character-long random string to be sent as a `challengeServer`.
+
+    **Note:** At the time of writing the PeerID Authentication specification does not contain recommendations about challenge length, but the official [`go-libp2p` implementation uses 32 characters](https://github.com/libp2p/go-libp2p/blob/master/p2p/http/auth/internal/handshake/handshake.go#L21).
+	3. Node generates `sig`, `headers` and `payload` as follows, where `peer-privkey` is the private key of the node's libp2p peer and `multiaddrs` is a list of string representations of the libp2p peer's multiaddresses:
+	```
+		sig = base64URL(
+			peer-privkey.sign(
+				bytes(varint + "challenge-node={challenge-node}") +
+				bytes(varint + "hostname={hostname}") +
+				bytes(varint + "server-public-key={public-key}")
+			)
+		)
+
+		headers = {
+			"Content-Type": "application/json",
+			"User-Agent": "some-user-agent",
+			"authorization": "libp2p-PeerID public-key=\"{nodePublicKeyB64}\", opaque=\"{opaque}\", challenge-server=\"{challengeServer}\", sig=\"{sig}\""
+		}
+
+        payload = {
+            "value": keyAuthorization,
+            "addresses": multiaddrs
+        }
+	```
+
+    **Note:** `varint` is a protobuf [varint](https://protobuf.dev/programming-guides/encoding/#varints) field that encodes the length of each of the `key=value` string.
+
+    **Note:** The node SHOULD include only multiaddresses containing public IPv4 addresses in `multiaddrs`.
```

Comment:
iirc there is more this (IPv6, filtering):

- The node MUST include only publicly reachable addresses
- For IPv4: exclude private ranges
- For IPv6: exclude private ranges and NAT64 translated addresses
- Exclude relay addresses (containing `/p2p-circuit`)